File: apps/mobile/lib/domain/models/tournament/tournament.dart
// Tournament Model — VSP Mobile App
//
// Domain model for a tournament event.
// Per Story 12.1 AC.
//
// Story 12.1 Slice F

import 'package:equatable/equatable.dart';

import 'tournament_format.dart';
import 'tournament_status.dart';

/// A tournament event — owns the tournament policy, players, flights, and results.
/// Per Story 12.1.
class Tournament extends Equatable {
  final String id;
  final String name;
  final TournamentFormat format;
  final TournamentStatus status;
  final int courseId;
  final String? courseName; // denormalized for display
  final DateTime startDate;
  final DateTime endDate;
  final String? tournamentPolicyId;
  final DateTime? registrationDeadline;
  final int? maxPlayers;
  final String? description;
  final DateTime createdAt;
  final int createdBy;
  final int version;

  /// Current leaderboard version — mobile can detect stale leaderboard.
  final int leaderboardVersion;

  const Tournament({
    required this.id,
    required this.name,
    required this.format,
    required this.status,
    required this.courseId,
    this.courseName,
    required this.startDate,
    required this.endDate,
    this.tournamentPolicyId,
    this.registrationDeadline,
    this.maxPlayers,
    this.description,
    required this.createdAt,
    required this.createdBy,
    this.version = 1,
    this.leaderboardVersion = 0,
  });

  /// True if this tournament is currently accepting registrations.
  bool get isRegistrationOpen =>
      status == TournamentStatus.registrationOpen &&
      (registrationDeadline == null ||
          registrationDeadline!.isAfter(DateTime.now()));

  /// True if the tournament is active (in progress).
  bool get isActive => status == TournamentStatus.inProgress;

  /// Parse from API JSON response.
  factory Tournament.fromJson(Map<String, dynamic> json) {
    return Tournament(
      id: json['id'] as String,
      name: json['name'] as String,
      format: TournamentFormatExtension.fromString(json['format'] as String),
      status: TournamentStatusExtension.fromString(json['status'] as String),
      courseId: json['courseId'] is String
          ? int.parse(json['courseId'] as String)
          : (json['courseId'] as num).toInt(),
      courseName: json['courseName'] as String?,
      startDate: DateTime.parse(json['startDate'] as String),
      endDate: DateTime.parse(json['endDate'] as String),
      tournamentPolicyId: json['tournamentPolicyId'] as String?,
      registrationDeadline: json['registrationDeadline'] != null
          ? DateTime.parse(json['registrationDeadline'] as String)
          : null,
      maxPlayers: (json['maxPlayers'] as num?)?.toInt(),
      description: json['description'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      createdBy: json['createdBy'] is String
          ? int.parse(json['createdBy'] as String)
          : (json['createdBy'] as num).toInt(),
      version: (json['version'] as num?)?.toInt() ?? 1,
      leaderboardVersion: (json['leaderboardVersion'] as num?)?.toInt() ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'format': format.name,
    'status': status.name,
    'courseId': courseId,
    'courseName': courseName,
    'startDate': startDate.toIso8601String(),
    'endDate': endDate.toIso8601String(),
    'tournamentPolicyId': tournamentPolicyId,
    'registrationDeadline': registrationDeadline?.toIso8601String(),
    'maxPlayers': maxPlayers,
    'description': description,
    'createdAt': createdAt.toIso8601String(),
    'createdBy': createdBy,
    'version': version,
    'leaderboardVersion': leaderboardVersion,
  };

  Tournament copyWith({
    String? id,
    String? name,
    TournamentFormat? format,
    TournamentStatus? status,
    int? courseId,
    String? courseName,
    DateTime? startDate,
    DateTime? endDate,
    String? tournamentPolicyId,
    DateTime? registrationDeadline,
    int? maxPlayers,
    String? description,
    DateTime? createdAt,
    int? createdBy,
    int? version,
    int? leaderboardVersion,
  }) {
    return Tournament(
      id: id ?? this.id,
      name: name ?? this.name,
      format: format ?? this.format,
      status: status ?? this.status,
      courseId: courseId ?? this.courseId,
      courseName: courseName ?? this.courseName,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      tournamentPolicyId: tournamentPolicyId ?? this.tournamentPolicyId,
      registrationDeadline: registrationDeadline ?? this.registrationDeadline,
      maxPlayers: maxPlayers ?? this.maxPlayers,
      description: description ?? this.description,
      createdAt: createdAt ?? this.createdAt,
      createdBy: createdBy ?? this.createdBy,
      version: version ?? this.version,
      leaderboardVersion: leaderboardVersion ?? this.leaderboardVersion,
    );
  }

  @override
  List<Object?> get props => [
    id,
    name,
    format,
    status,
    courseId,
    courseName,
    startDate,
    endDate,
    tournamentPolicyId,
    registrationDeadline,
    maxPlayers,
    description,
    createdAt,
    createdBy,
    version,
    leaderboardVersion,
  ];
}

File: apps/api/src/main/java/vnpt/vsp/module/tournament/entity/Tournament.java
package vnpt.vsp.module.tournament.entity;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Tournament entity — the main tournament aggregate root.
 * Per Story 12.1 AC: supports required formats, registration/import, flights,
 * tee times, starting tees, score confirmation, tie-break, and result publication.
 *
 * Tournament owns TournamentPolicy; individual rounds inherit it.
 */
@Entity
@Table(name = "tournaments")
public class Tournament {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "name", nullable = false)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(name = "format", nullable = false)
    private TournamentFormat format;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private TournamentStatus status = TournamentStatus.DRAFT;

    @Column(name = "course_id", nullable = false)
    private Long courseId;

    @Column(name = "start_date", nullable = false)
    private Instant startDate;

    @Column(name = "end_date", nullable = false)
    private Instant endDate;

        /**
     * This outing's rules as JSON — divisions, handicap cap, judging floor,
     * countback windows, daily-CAP scale, technical prizes.
     *
     * Per event, because they change per event: the club's own sheets already
     * disagree about where nhóm A ends, and the previous outing ran a different
     * number of flights. Null until an organiser configures them.
     */
    @Column(name = "outing_rules", columnDefinition = "jsonb")
    @org.hibernate.annotations.JdbcTypeCode(org.hibernate.type.SqlTypes.JSON)
    private String outingRules;

@Column(name = "tournament_policy_id")
    private UUID tournamentPolicyId;

    @Column(name = "registration_deadline")
    private Instant registrationDeadline;

    @Column(name = "max_players")
    private Integer maxPlayers;

    @Column(name = "description", length = 2000)
    private String description;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "created_by", nullable = false)
    private Long createdBy;

    @Version
    @Column(name = "version", nullable = false)
    private int version = 1;

    @OneToMany(mappedBy = "tournament", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TournamentPlayer> players = new ArrayList<>();

    @OneToMany(mappedBy = "tournament", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Flight> flights = new ArrayList<>();

    @OneToMany(mappedBy = "tournament", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TeeTime> teeTimes = new ArrayList<>();

    @OneToMany(mappedBy = "tournament", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TieBreakRule> tieBreakRules = new ArrayList<>();

    @OneToMany(mappedBy = "tournament", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TournamentResult> results = new ArrayList<>();

    /**
     * Leaderboard version — incremented on each score update.
     * Mobile clients can detect stale leaderboard by comparing version.
     */
    @Column(name = "leaderboard_version", nullable = false)
    private long leaderboardVersion = 0L;

    @PrePersist
    protected void onCreate() {
        createdAt = Instant.now();
    }

    // ─── Status transitions ───────────────────────────────────────────────────

    public boolean canOpenRegistration() {
        return status == TournamentStatus.DRAFT;
    }

    public boolean canStart() {
        return status == TournamentStatus.REGISTRATION_OPEN && !players.isEmpty();
    }

    public boolean canComplete() {
        return status == TournamentStatus.IN_PROGRESS;
    }

    public boolean isModifiable() {
        return status == TournamentStatus.DRAFT || status == TournamentStatus.REGISTRATION_OPEN;
    }

    // ─── Getters / Setters ───────────────────────────────────────────────────

    public UUID getId() { return id; }
    public void setId(UUID id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public TournamentFormat getFormat() { return format; }
    public void setFormat(TournamentFormat format) { this.format = format; }

    public TournamentStatus getStatus() { return status; }
    public void setStatus(TournamentStatus status) { this.status = status; }

    public Long getCourseId() { return courseId; }
    public void setCourseId(Long courseId) { this.courseId = courseId; }

    public Instant getStartDate() { return startDate; }
    public void setStartDate(Instant startDate) { this.startDate = startDate; }

    public Instant getEndDate() { return endDate; }
    public void setEndDate(Instant endDate) { this.endDate = endDate; }

    public UUID getTournamentPolicyId() { return tournamentPolicyId; }
    public void setTournamentPolicyId(UUID tournamentPolicyId) { this.tournamentPolicyId = tournamentPolicyId; }

    public Instant getRegistrationDeadline() { return registrationDeadline; }
    public void setRegistrationDeadline(Instant registrationDeadline) { this.registrationDeadline = registrationDeadline; }

    public Integer getMaxPlayers() { return maxPlayers; }
    public void setMaxPlayers(Integer maxPlayers) { this.maxPlayers = maxPlayers; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public Instant getCreatedAt() { return createdAt; }
    public Long getCreatedBy() { return createdBy; }
    public void setCreatedBy(Long createdBy) { this.createdBy = createdBy; }

    public int getVersion() { return version; }
    public void setVersion(int version) { this.version = version; }

    public List<TournamentPlayer> getPlayers() { return players; }
    public void setPlayers(List<TournamentPlayer> players) { this.players = players; }

    public List<Flight> getFlights() { return flights; }
    public void setFlights(List<Flight> flights) { this.flights = flights; }

    public List<TeeTime> getTeeTimes() { return teeTimes; }
    public void setTeeTimes(List<TeeTime> teeTimes) { this.teeTimes = teeTimes; }

    public List<TieBreakRule> getTieBreakRules() { return tieBreakRules; }
    public void setTieBreakRules(List<TieBreakRule> tieBreakRules) { this.tieBreakRules = tieBreakRules; }

    public List<TournamentResult> getResults() { return results; }
    public void setResults(List<TournamentResult> results) { this.results = results; }

    public long getLeaderboardVersion() { return leaderboardVersion; }
    public void setLeaderboardVersion(long leaderboardVersion) { this.leaderboardVersion = leaderboardVersion; }

    public void incrementLeaderboardVersion() {
        this.leaderboardVersion++;
    }

    public String getOutingRules() { return outingRules; }
    public void setOutingRules(String outingRules) { this.outingRules = outingRules; }
}
